ansible/roles/csm.ncn-initrd/tasks/main.yml

```diff
@@ -0,0 +1,29 @@
+#
+# MIT License
+#
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+---
+# Create an initrd for the NCN image being modified by CFS
+
+- name: Create NCN initrd
+  command: /srv/cray/scripts/common/create-ims-initrd.sh
```

Comment:
Is this just calling dracut? If so, we should probably just use mkinitrd with proper configuration files in place.

Is the intention for every NCN to create an initrd, or can we create it once on one NCN and use it for all of them?

Reply:
That script invokes dracut with specific options, removes the kdump initrd, and copies files into place to support booting from disk vs pxe.  https://github.com/Cray-HPE/metal-provision/blob/main/roles/ncn-common/files/scripts/common/create-ims-initrd.sh

The initrd is not the same across all NCNs...

Reply:
In general we create a non host only initrd once during the NCN build per Ncn type, one initrd for K8s and another for storage-ceph (they have different dracut modules).

The dracut config we have only sets a few common configurations, the rest are specified on the cmdline when dracut is invoked.

This ims script is mimicking what we do in general, but in the context of ims instead of during the Ncn build.

I suppose we could use mkinitrd, but off hand I'm not sure if we're doing anything that wouldn't be covered.

Reply:
I think its up to you guys re: mkinitrd; personally having to know the full set of dracut options to include each time a build is made can make for overall maintaining and generating new initrds painful, unless you use a custom one-off script like this (and advertise it to people who are recompiling and building things by hand).

I know that some dracut modules pull from system provided filesystems that IMS may be putting in place in order to construct the image root. This can have a few different effects, but the concern would be that you might get a flavor of initrd based on what node is hosting the image pod during build time.

The other issue, is that it might eventually make sense to generate BOS session templates for NCN nodes, and if all nodes truly did have a requirement for a unique initrd, it would be much more headache to maintain (using the 'register all artifacts to s3' approach that computes uses).

Reply:
The only requirement we have for different intirds is between k8s and ceph.
When we rebuild the initrd in IMS with slingshot drivers that initrd could be assigned to all k8s nodes but we only set it for worker nodes, so at that point we have 3 different initrds instead of 2.

By default our configuration uses hostonly to facilitate keeping initrds small (such as kdumps), and then our one-off scripts override that with --no-host-only. I believe the --no-host-only would circumvent the issue you pointed out about the host nodes loaded modules.. but maybe I misunderstood which aspect you were speaking to.

Reply:
Are there multiple types of computes?

NCNs have 3 types of nodes, which at worst case each type has it's own initrd. The bos template sounds like it would register per type, which is easy for the single focused, diskless computes?